lib/Basics/debugging.h

```diff
@@ -199,31 +200,51 @@ enable_if_t<is_container<T>::value, std::ostream&> operator<<(std::ostream& o, T
   return o;
 }
 
+namespace debug {
+
+struct NoOpStream {
+  template <typename T>
+  auto operator<<(T const&) -> NoOpStream& {
+    return *this;
+  }
+};
+
+struct AssertionLogger {
+  void operator&(std::stringstream const& str) {
+    std::string message = str.str();
+    arangodb::CrashHandler::assertionFailure(file, line, function, expr,
+                                             message.empty() ? nullptr : message.c_str());
+  }
+  void operator&(std::ostream const& os) {
+    operator&(static_cast<std::stringstream const&>(os));
+  }
+
+  const char* file;
+  int line;
+  const char *function;
+  const char* expr;
+};
+
+}  // namespace debug
+
 }  // namespace arangodb
 
 /// @brief assert
 #ifndef TRI_ASSERT
 
 #ifdef ARANGODB_ENABLE_MAINTAINER_MODE
 
-#define TRI_ASSERT(expr) /*GCOVR_EXCL_LINE*/                                             \
-  do {                                                                                   \
-    if (!(ADB_LIKELY(expr))) {                                                           \
-      arangodb::CrashHandler::assertionFailure(__FILE__, __LINE__, __FUNCTION__, #expr); \
-    } else {                                                                             \
-    }                                                                                    \
-  } while (false)
+#define TRI_ASSERT(expr) /*GCOVR_EXCL_LINE*/                                        \
+  (!(ADB_LIKELY(expr)))                                                             \
+      ? (void)nullptr                                                               \
+      : arangodb::debug::AssertionLogger{__FILE__, __LINE__, __FUNCTION__, #expr} & \
+            (std::stringstream{})
```

Comment:
Shouldn't this be the other way around, i.e. if 
```suggestion
  (ADB_LIKELY(expr))                                                                \
      ? (void)nullptr                                                               \
      : arangodb::debug::AssertionLogger{__FILE__, __LINE__, __FUNCTION__, #expr} & \
            (std::stringstream{})
```